### PR TITLE
Fix df.greatest()

### PR DESCRIPTION
**Pull request by zurk**

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@zurk tests do not pass (as they should I guess) with

```
AssertionError: {'993': 500, '804': 500, '810': 500, '851': 500, [5946 chars] 602} != {'993': 993, '804': 804, '810': 810, '851': 851, [5946 chars] 923}
```

Weird - all frequencies set to 500.

**Comment by zurk** [comment]
No, they should not, because it was a just new test without code change. It was how old `df.greatest()` works.

File: sourced/ml/models/df.py [hunk 1]
```diff
@@ -103,31 +103,26 @@ def greatest(self, max_size: int) -> "DocumentFrequencies":
...
-                border_keys.append(key)
-            chosen_pos += 1
-            if chosen_pos >= len(chosen):
-                break
-            chosen_val = chosen[chosen_pos]
-        border_keys.sort()
-        for k in border_keys[:max_size - len(df)]:
-            df[k] = border_freq
+        if max_size != freqs.shape[0]:
+            border_freq_indexes = freqs == border_freq
```
Comment (vmarkovtsev) [inline sourced/ml/models/df.py]: `assert max_size < freqs.shape[0]`

File: sourced/ml/models/df.py [hunk 2]
```diff
@@ -103,31 +103,26 @@ def greatest(self, max_size: int) -> "DocumentFrequencies":
         self._log.info("Pruning to max %d size", max_size)
         pruned = type(self)()
         pruned._docs = self.docs
-        freqs = numpy.fromiter(self._df.values(), numpy.int32, len(self))
+        keys, freqs = list(zip(*self._df.items()))
+        freqs = numpy.array(freqs, numpy.int32)
+        keys = numpy.array(keys)
```
[the remark this answers is not in the document]
Comment (zurk) [reply sourced/ml/models/df.py]: discussed, we decide to use `obj_array = np.array(l, dtype=object)` it does not have this memory problem.

**Comment by zurk** [comment]
I do `keys = numpy.array(list(self._df.keys()), dtype=object)` because `keys = numpy.fromiter(self._df.keys(), dtype=object)` unavaliable for `fromiter` function. Hope it is not a big deal.

File: sourced/ml/models/df.py [hunk 3]
```diff
@@ -103,31 +103,28 @@ def greatest(self, max_size: int) -> "DocumentFrequencies":
...
-            chosen_val = chosen[chosen_pos]
-        border_keys.sort()
-        for k in border_keys[:max_size - len(df)]:
-            df[k] = border_freq
+        if max_size != freqs.shape[0]:
+            assert max_size < freqs.shape[0]
+            border_freq_indexes = freqs == border_freq
+            border_keys = keys[border_freq_indexes]
+            border_keys.sort()
+            border_keys = border_keys[:len(border_keys) + max_size - freqs.shape[0]]
```
Comment (vmarkovtsev) [inline sourced/ml/models/df.py]: do we need `len(border_keys)` here? the negative value should work

Comment (zurk) [reply sourced/ml/models/df.py]: Indeed

**Pull request merged**